_posts/2021-09-14-decathlon-user-story.adoc

```diff
@@ -0,0 +1,207 @@
+
+Finally, to end my feedback, I regret not having faced the construction of a native image. But our project has opted for the deployment of containers within a Kubernetes cluster, which is what Quarkus is designed for.
+
+=== __Loïc Mathieu - Senior Developer__
+
+I’m a regular Quarkus committer and knows very well Quarkus before I joined the team. So I will give a very narrow testimonial, on a very specific feature we used and that I very love.
+
+Reactive programming is at the heart of Quarkus, and reactive messaging, as its name implies, is a reactive framework.
+
+When you need to consume or produce a message, you can simply use the payload as method parameter or return type. But when you need to implement asynchronous processing or complex logic on a stream of messages, you’ll need to use Mutiny instead of directly using your payload type.
+Mutiny is a set of reactive types and operators. It allows to express a set of transformations on a stream of items and  follows the reactive stream standard.
+
+One of our needs was to group incoming messages into batches, as sending a message to an ERP module has a cost, and they fear receiving millions of messages per day. So we need to group them by type, then by batch of 50, and emit at least one batch per minute to avoid adding too much delay to the message delivery.
+
+To implement this, we prototype the usage of Kafka Stream, but it adds some complexity to our current technical stack as we didn’t use it already, and it has some grouping limitations (grouping by size is not provided out of the box by the framework).
+
+So we decided to simply use what we already have in our toolbox, Mutiny, and the code we implement is really readable for such a complex task, and we have been very happy with it since.
+
+[code, java]
+----
+@Incoming("in")
+@Outgoing("group-out")
+public Multi<Message<List<ErpMessage>>> group(Multi<KafkaRecord<String, RawMessage>> events) { // <1>
+    return events
+            .group().by(record -> record.getPayload().type) // <2>
+            .flatMap(group -> group.group().intoLists().of(size, duration)) // <3>
+            .filter(group -> !group.isEmpty()) // <4>
+            .flatMap(groupedMultis -> {
+                List<ErpMessage> erpMsg = groupedMultis.stream() // <5>
+                        .map(record -> toErpMessage(record))
+                        .collect(Collectors.toList());
+                 // <6>
+                return Multi.createFrom().item(KafkaRecord.of((String) null, erpMsg));
+            });
+}
+----
+
+1. The method takes a stream of Kafka messages.
+2. First, we group by payload type.
+3. Then, we group by batch of size messages with a max duration of duration.
+4. Then, we remove empty batches.
+5. Then, we map each batch of message to a new format.
+6. Then we return a Kafka message with the batch in it.
```

Comment:
The <6> anchor in the code is misplaced. It should be at the end of a line.

Reply:
Fixed, thanks